packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Models/Types/ModelTypeProvider.cs

```diff
@@ -121,5 +121,203 @@ private bool PropertyHasSetter(CSharpType type, InputModelProperty prop)
+                : _inputModel.Usage.HasFlag(InputModelTypeUsage.Input)
+                    ? MethodSignatureModifiers.Public
+                    : MethodSignatureModifiers.Internal;
+            var constructorParameters = BuildConstructorParameters(false);
+
+            var constructor = new CSharpMethod(
+                signature: new ConstructorSignature(
+                    Type,
+                    $"Initializes a new instance of {Type:C}",
+                    null,
+                    accessibility,
+                    constructorParameters),
+                body: new MethodBodyStatement[]
+                {
+                    new ParameterValidationBlock(constructorParameters),
+                    GetPropertyInitializers(constructorParameters)
+                },
+                kind: CSharpMethodKinds.Constructor);
+
+            return constructor;
+        }
+
+        private CSharpMethod? BuildSerializationConstructor()
+        {
+            // The property bag never needs deserialization, therefore we return the null here
+            if (_inputModel.IsPropertyBag)
+                return null;
+
+            var constructorParameters = BuildConstructorParameters(true);
+
+            return new CSharpMethod(
+                signature: new ConstructorSignature(
+                    Type,
+                    $"Initializes a new instance of {Type:C}",
+                    null,
+                    MethodSignatureModifiers.Internal,
+                    constructorParameters),
+                body: new MethodBodyStatement[]
+                {
+                    new ParameterValidationBlock(constructorParameters),
+                    GetPropertyInitializers(constructorParameters)
+                },
+                kind: CSharpMethodKinds.Constructor);
+        }
+
+        private MethodBodyStatement[] GetPropertyInitializers(IReadOnlyList<Parameter> parameters)
+        {
+            List<MethodBodyStatement> methodBodyStatements = new();
+
+            Dictionary<string, Parameter> parameterMap = parameters.ToDictionary(
+                parameter => parameter.Name,
+                parameter => parameter);
+
+            foreach (var property in Properties)
+            {
+                ValueExpression? initializationValue = null;
+
+                if (parameterMap.TryGetValue(property.Name.ToVariableName(), out var parameter) || IsStruct)
+                {
+                    if (parameter != null)
+                    {
+                        initializationValue = new ParameterReference(parameter);
+
+                        if (CSharpType.RequiresToList(parameter.Type, property.Type))
+                        {
+                            initializationValue = parameter.Type.IsNullable ?
+                                Linq.ToList(new NullConditionalExpression(initializationValue)) :
+                                Linq.ToList(initializationValue);
```

Comment:
I think this could be simplified since `initializationValue` is a typed expression:
```suggestion
                            initializationValue = Linq.ToList(initializationValue.NullConditional());
```

Reply:
This won't work, since initializationValue is a ValueExpression and I can't assign it to typed expression because, it is used to initialize new instance as well in case of collection .

Reply:
`initializationValue` is `ValueExpression` because you defined in this way.
The instance here we are using is `ParameterReference` which inherits `TypedValueExpression`.
For example, it is totally fine if we do this:
```
var parameterVar = new ParameterReference(parameter);
if (CSharpType.RequiresToList(parameter.Type, property.Type))
{
	initializationValue = Linq.ToList(parameterVar.NullConditional());
}
else
{
	initializationValue = parameterVar;
}
```
of cource we could change this again to a ternary operator expression to simplify it further more